packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1327 @@
+  /// that the [build] method uses before they complete their respective
+  /// futures.
+  ///
+  /// Typically this method returns a suitably-configured [Navigator]. If you do
+  /// plan to create a navigator, consider using the
+  /// [PopNavigatorRouterDelegateMixin].
+  ///
+  /// This method must not return null.
+  ///
+  /// The `context` is the [Router]'s build context.
+  Widget build(BuildContext context);
+}
+
+/// A route information provider that provides route information for the
+/// [Router] widget
+///
+/// This provider is responsible for handing the route information through [value]
+/// getter and notifies listeners, typically the [Router] widget, when a new
+/// route information is available.
+///
+/// If you decide to opt for the route information reporting (by overrides the
+/// [RouterDelegate.currentConfiguration] to return non-null), overrides the
+/// [routerReportsNewRouteInformation] method to process the route information.
+///
+/// See also:
+///
+///  * [PlatformRouteInformationProvider], which wires up the provider with the
+///    [WidgetsBindingObserver.didPushRoute] to propagate platform push route
+///    intent to the [Router] widget, as well as reports new route information
+///    from the [Router] back to the engine by overriding the
+///    [routerReportsNewRouteInformation].
+abstract class RouteInformationProvider
+    extends ValueListenable<RouteInformation> {
+  /// A callback called when the [Router] widget detects any navigation event
+  /// due to state changes.
+  ///
+  /// The subclasses can override this method to update theirs values or trigger
+  /// other side effects. For example, the [PlatformRouteInformationProvider]
+  /// overrides this method to report the route information back to the engine.
+  ///
+  /// The [routeInformation] contains the route information after the
+  /// navigation event.
+  void routerReportsNewRouteInformation(RouteInformation routeInformation) {}
+}
+
+/// The route information provider that propagates the platform route information changes.
+///
+/// This provider also reports the new route information from the [Router] widget
+/// back to engine using message channel method, the
+/// [RouteNotificationMessages.maybeNotifyRouteChange].
+class PlatformRouteInformationProvider extends RouteInformationProvider
+    with WidgetsBindingObserver, ChangeNotifier {
+  /// Create a platform route information provider.
+  ///
+  /// Use the [initialRouteInformation] to set the default route information for this
+  /// provider.
+  PlatformRouteInformationProvider({RouteInformation initialRouteInformation})
+      : _value = initialRouteInformation,
+        _currentSerializedRouteInformationInEngine = _serializeRouteInformation(initialRouteInformation);
+
+  String _currentSerializedRouteInformationInEngine;
+
+  static String _serializeRouteInformation(RouteInformation routeInformation) {
```

Comment:
This isn't the right location to serialize/deserialize. The serialization should happen in `RouteNotificationMessages.maybeNotifyRouteChange` just before things are put on the channel. And the deserialization of the engine information should happen in whoever calls our WidgetsBindingObserver implementation.